src/process-events.c

```diff
@@ -400,526 +349,289 @@ static __always_inline int write_path(struct pt_regs *ctx, void *ptr, buf_t *buf
-    {
-        // if this message failed to be sent let's remove the events
-        // from the maps so the kretprobe doesn't do useless work
-        u64 pid_tgid = bpf_get_current_pid_tgid();
-        u32 pid = pid_tgid >> 32;
-        bpf_map_delete_elem(&exec_tids, &pid);
-        bpf_map_delete_elem(&incomplete_execs, &pid_tgid);
-    }
+    // do not emit if we couldn't fill the syscall info
+    if (fill_syscall(&pm->u.syscall_info, ts, pid_tgid >> 32) != 0) goto Done;
 
-    return 0;
-}
+    void *mmptr = is_user_process(ts);
+    if (mmptr == NULL) goto Done; // this is OK; it just means not a user process
 
-static __always_inline int write_pwd(struct pt_regs *ctx, buf_t *buffer, u32 *offset,
-                                     u32 *skips, tail_call_slot_t tc_slot, error_info_t *einfo)
-{
-    void *ts = (void *)bpf_get_current_task();
-    int ret = -PMW_READING_FIELD;
-
-    void *pwd_ptr = NULL;
-    // task_struct->fs
-    if (read_value(ts, CRC_TASK_STRUCT_FS, &pwd_ptr, sizeof(pwd_ptr)) < 0)
+    void *exe = NULL;
+    read_value(mmptr, CRC_MM_STRUCT_EXE_FILE, &exe, sizeof(exe));
+    if (!exe) // this is wholly unexpected so emit a warning
     {
-        einfo->offset_crc = CRC_TASK_STRUCT_FS;
+        ret = -PMW_MISSING_EXE;
         goto Done;
     }
 
-    // &(fs->pwd)
-    pwd_ptr = offset_ptr(pwd_ptr, CRC_FS_STRUCT_PWD);
-    if (pwd_ptr == NULL)
-    {
-        einfo->offset_crc = CRC_FS_STRUCT_PWD;
+    u64 arg_start = 0;
+    u64 arg_end = 0;
+    read_value(mmptr, CRC_MM_STRUCT_ARG_START, &arg_start, sizeof(arg_start));
+    read_value(mmptr, CRC_MM_STRUCT_ARG_END, &arg_end, sizeof(arg_end));
+    if (arg_end < arg_start) {
+        ret = -PMW_ARGV_INCONSISTENT;
+        pm->u.warning_info.info.argv.start = arg_start;
+        pm->u.warning_info.info.argv.end = arg_end;
         goto Done;
     }
 
-    ret = write_path(ctx, pwd_ptr, buffer, skips, offset, tc_slot, einfo);
-    if (ret < 0)
-        goto Done;
-
-    ret = 0;
-
-    // add an extra null byte to signify string section end
-    write_null_char(buffer, offset);
-
-Done:;
-    // reset skips back to 0. This will automatically update it in the
-    // map so no need to do a bpf_map_update_elem.
-    *skips = 0;
-    return ret;
-}
+    // length of all strings counting NULLs
+    u64 argv_len = arg_end - arg_start;
+    const u32 MAX_ARGV_LEN = (MAX_PERCPU_BUFFER >> 1) - 1;
+    if (argv_len > MAX_ARGV_LEN) {
+        // deliberately making a new process_message_t instead of
+        // reusing `pm`. The goal is to still push the truncated argv
+        // so we do not want to override any of the data in `pm`
+        process_message_t warning = {0};
+        warning.type = PM_WARNING;
+        warning.u.warning_info.pid_tgid = pid_tgid;
+        warning.u.warning_info.message_type = pm_type;
+        warning.u.warning_info.code = PMW_ARGV_TOO_BIG;
+        warning.u.warning_info.info.total_len = argv_len;
+        push_message(ctx, &warning);
```

Comment:
Is there maybe room somewhere in `pm` to store a "truncated" flag instead of pushing a whole new message?

Reply:
If we did that then I would want it to be a truncated and an original length field so we know that on the cases this may not be enough how much more it may need to be.

Reply:
Done! (: